### DISCUSSION TITLE
How to create connection in an explicit m-n relationship?

### DISCUSSION THREAD

**Discussion by GeoffreyDick** [prisma/prisma#4339] [Q&A] [answered]
Currently stuck on how to create a connection with an explicit m-n relationship. I've defined the Category/Item relationship explicitly because I require a Quantity to be attached to the relationship. I'm using the Prisma Client on NestJS with GraphQL. 

My Prisma schema is basically just a copy-paste of the [Explicit m-n documentation](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#many-to-many-relations) with a few extra fields:
```prisma
// schema.prisma
model Category {
  id        String         @id @default(uuid())
  createdAt DateTime       @default(now())
  updatedAt DateTime       @updatedAt
  name      String
  items     CategoryItem[]
  list      List           @relation(fields: [listId], references: [id])
  listId    String
}

model CategoryItem {
  category   Category @relation(fields: [categoryId], references: [id])
  categoryId String
  item       Item     @relation(fields: [itemId], references: [id])
  itemId     String
  quantity   Int      @default(1)
  createdAt  DateTime @default(now())
  updatedAt  DateTime @updatedAt

  @@id([categoryId, itemId])
}

model Item {
  id         String         @id @default(uuid())
  createdAt  DateTime       @default(now())
  updatedAt  DateTime       @updatedAt
  brand      String?
  model      String
  weight     Float?
  categories CategoryItem[]
  user       User           @relation(fields: [userId], references: [id])
  userId     String
}
```

This is where I'm stuck. Based on the [Nested Writes documentation](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes) and #3266, my understanding is that there is currently no way to perform a nested write with an explicit m-n relationship. So I tried the following solution: create the item first, and then update the category to connect the newly created item. Not the most efficient but it should work, right? Didn't work. I also tried: create the item first, and then update the item, and that also didn't work. Both times I received the same error. Here's my NestJS resolver:

```typescript
// items.resolver.ts
  @Mutation((returns) => Item)
  async createItem(@Args('data') data: ItemInput) {
    const item = await this.prisma.item.create({
      data: {
        brand: data.brand,
        model: data.model,
        weight: data.weight,
        userId: data.userId,
      },
    });

    return await this.prisma.category.update({
      where: {
        id: data.categoryId,
      },
      data: {
        items: {
          connect: {
            categoryId_itemId: {
              categoryId: data.categoryId,
              itemId: item.id,
            },
          },
        },
      },
    });
  }
```

Here's my GraphQL mutation:
```graphql
mutation {
  createItem(data:{
    brand: "Acme"
    model: "TNT"
    weight: 1000
    userId: "e4f6a563-b77c-4759-90b5-9ce278f42036"
    categoryId: "9cd76518-1da2-46b7-a591-5c1a64406f30"
  }) {
    id
    brand
    model
    weight
    categories {
      quantity
      category {
        name
      }
    }
  }
}
```
And here is the error: 
```json
{
  "errors": [
    {
      "message": "\nInvalid `prisma.category.update()` invocation:\n\n\n  The required connected records were not found. Expected 1 records to be connected, found 0.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "createItem"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "code": "P2018",
          "clientVersion": "2.11.0",
          "meta": {
            "details": "Expected 1 records to be connected, found 0."
          },
          "stacktrace": [
            "Error: ",
            "Invalid `prisma.category.update()` invocation:",
            "",
            "",
            "  The required connected records were not found. Expected 1 records to be connected, found 0.",
            "    at PrismaClientFetcher.request (/project/path/node_modules/@prisma/client/runtime/index.js:78782:15)",
            "    at processTicksAndRejections (node:internal/process/task_queues:93:5)",
            "    at async ItemsResolver.createItem (/project/path/dist/src/items/items.resolver.js:39:16)",
            "    at async target (/project/path/node_modules/@nestjs/core/helpers/external-context-creator.js:76:28)",
            "    at async /project/path/node_modules/@nestjs/core/helpers/external-proxy.js:9:24"
          ]
        }
      }
    }
  ],
  "data": null
}
```
When I run the mutation, the item is successfully created in the database, so referencing `item.id` should work in the `category.update()` function. And the `categoryId` is copied directly from my database so both should be referenced correctly. Is there a different way I should be doing this? I couldn't find anything in the documentation.

Thanks in advance!

**Comment by ryands17** [comment, marked as answer]
Hey @GeoffreyDick 👋 

> This is where I'm stuck. Based on the Nested Writes documentation and #3266, my understanding is that there is currently no way to perform a nested write with an explicit m-n relationship

There is a way but you won't get the response in the same way that you do with an implicit m-n relationship.

You can perform nested writes in the following manner for explicit m-n relationships:

```ts
await prisma.category.create({
    data: {
      name: 'category 1',
      // join table
      items: {
        create: [
          { quantity: 10, item: { create: { model: 'item-1' } } },
          { quantity: 20, item: { create: { model: 'item-2' } } },
        ],
      },
    },
})
```

This creates the `category`, the join table and the `item` as well.

As for this query:

```ts
const item = await this.prisma.item.create({
      data: {
        brand: data.brand,
        model: data.model,
        weight: data.weight,
        userId: data.userId,
      },
    });

return await this.prisma.category.update({
      where: {
        id: data.categoryId,
      },
      data: {
        items: {
          connect: {
            categoryId_itemId: {
              categoryId: data.categoryId,
              itemId: item.id,
            },
          },
        },
      },
});
```

I'm assuming you want to add an item to that category right? If that's the case, you can simply do this:

```ts
await prisma.category.update({
    where: { id: 'something' },
    data: {
      items: {
        create: { item: { create: { model: 'item 1' } } },
      },
    },
})
```

This will create a relation in the join table mapping the category to this item. Let me know if I've misunderstood anything here :)

**Reply by GeoffreyDick** [reply]
Ahh ok that last example is what I needed. I tried that previously but without the second `create:`. My new mutation is now:
```typescript
  @Mutation((returns) => Item)
  async createItem(
    @Args('data') data: ItemInput,
    @Args('categoryId') categoryId: string,
  ) {
    return this.prisma.item.create({
      data: {
        ...data,
        categories: {
          create: {
            quantity: 1,
            category: {
              connect: {
                id: categoryId,
              },
            },
          },
        },
      },
    });
```
Thanks for the help!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
